### ISSUE TITLE
[Bug]: `with control(q0, q1): z(target)` (multi-controlled Z, 2+ controls) produces a wrong unitary with spurious imaginary components

### ISSUE THREAD

**Issue by kkoci** [Quantinuum/guppylang#2251]
### Description

A multi-controlled Z gate built with 2 or more control qubits via the `control` modifier (`with control(q0, q1): z(target)`) produces a unitary that is NOT the CCZ gate, applied to a uniform superposition, two of the eight basis-state amplitudes come back with spurious *imaginary* components instead of the expected real +/-1 phase flip, and probability is redistributed onto the imaginary axis rather than remaining a pure phase flip. 

The same `control` modifier with an `x` target (`with control(q0, q1): x(target)`) is correct and matches the built-in `toffoli` gate exactly, so this looks specific to how a Z-type (rather than X-type) target is synthesized under multi-qubit control. A workaround exists,sandwich the target with Hadamards and use multi-controlled X instead (`h(t); with control(...): x(t); h(t)`) confirmed exact to ~1e-16 against the true CCZ/CCCZ unitary for both 2 and 3 controls, and is used throughout our project.

### Steps to Reproduce

1. Allocate 3 qubits and put all of them in |+++> with Hadamard gates.
2. Apply `with control(q0, q1): z(q2)` (a 2-controlled Z targeting q2).
3. Read out the full statevector via `state_output` + `.emulator().statevector_sim().run()` + `partial_state_dicts()`.
4. Compare against the exact CCZ|+++> result: uniform +0.35355 (1/sqrt(8)) on all 8 basis states except |111>, which should be -0.35355  everything real, only index 7 negated.
5. Separately, repeat with `with control(q0, q1): x(q2)` sandwiched between `h(q2)` before and after (the standard H-CCX-H identity) and observe it reproduces the correct result exactly.
6. Minimal script reproducing both cases attached in Additional Context below.

### Expected Behaviour

`with control(q0, q1): z(q2)` applied to |+++> should produce the exact CCZ unitary: every amplitude real and equal to +1/sqrt(8) (0.35355...), except the |111> (index 7) amplitude, which should be -1/sqrt(8). No amplitude should acquire an imaginary component  Z, H, and control are all real gates, so the whole circuit's output must be real-valued.

### Actual Behaviour

Two amplitudes (indices 6 and 7) come back with spurious imaginary components instead of the expected real values, and the phase flip lands on the wrong structure (a rotation into the imaginary axis on two states, rather than a real sign flip on one). The H·CCX·H workaround, run alongside for comparison, is exact.

| Index | control(q0,q1): z(q2) (actual, buggy) | H·control(q0,q1):x(q2)·H (workaround) | Expected CCZ\|+++> |
|---|---|---|---|
| 0 | 0.3536 + 0j | 0.3536 + 0j | 0.3536 + 0j |
| 1 | 0.3536 + 0j | 0.3536 + 0j | 0.3536 + 0j |
| 2 | 0.3536 + 0j | 0.3536 + 0j | 0.3536 + 0j |
| 3 | 0.3536 + 0j | 0.3536 + 0j | 0.3536 + 0j |
| 4 | 0.3536 + 0j | 0.3536 + 0j | 0.3536 + 0j |
| 5 | 0.3536 + 0j | 0.3536 + 0j | 0.3536 + 0j |
| 6 | 0 − 0.3536j | 0.3536 + 0j | 0.3536 + 0j |
| 7 | 0 + 0.3536j | −0.3536 + 0j | −0.3536 + 0j |

Max abs diff, buggy result vs expected: 0.3536 (a full amplitude's worth). Max abs diff, workaround vs expected: 2.7755575615628914e-16.

Reproduced identically for 3 controls (4-qubit register, CCCZ on |1111>): direct 3-control Z gives max abs diff vs expected = 0.35355339059327373; H·(3-control X)·H workaround gives max abs diff vs expected = 2.7755575615628914e-16.

Circuit still compiles and runs without error or warning in both the 2- and 3-control cases the wrong result is silent.

### Component

Guppy Compiler

### Environment

```markdown
guppylang: 1.0.2
Python: 3.14.5
OS: Windows 11 (10.0.26200)
hugr: 0.18.5
selene-sim: 0.3.0
selene-hugr-qis-compiler: 0.4.2
```

### Additional Context

This is a separate, unrelated bug from #2250 different symptom and different code path. #2250 is about `iqft` (built by daggering/reversing another comptime-generic function) producing wrong results specifically when run in isolation vs. combined with its counterpart in the same program a monomorphization/state-sharing issue with no connection to gate synthesis. This issue is about the `control` modifier itself synthesizing an incorrect unitary for a Z-type target under 2+ qubit control, independent of comptime it reproduces identically in a single, minimal, non-generic program with nothing else going on.

Full writeup, including the 3-control confirmation and what we ruled out (this is not a `dagger` or `comptime` interaction plain X-type targets under the same multi-qubit `control` modifier are correct and match `toffoli` exactly), is in CLAUDE.md (the "grover" section, gotcha #12) of kkoci/Qshelf (https://github.com/kkoci/Qshelf). The originating regression test is packages/grover/tests/test_correctness.py::test_direct_multi_controlled_z_is_broken (kept as xfail(strict=True)); the production workaround is used throughout packages/grover/src/grover/grover.py.

Minimal repro (both the broken case and the working workaround, for comparison):

```
import numpy as np
from guppylang import guppy
from guppylang.std.builtins import array, control
from guppylang.std.debug import state_output
from guppylang.std.quantum import qubit, h, x, z, discard_array

@guppy  # broken: direct multi-controlled Z
def broken() -> None:
    qs = array(qubit() for _ in range(3))
    h(qs[0]); h(qs[1]); h(qs[2])
    with control(qs[0], qs[1]):
        z(qs[2])
    state_output("out", qs)
    discard_array(qs)

@guppy  # correct: H . multi-controlled-X . H workaround
def workaround() -> None:
    qs = array(qubit() for _ in range(3))
    h(qs[0]); h(qs[1]); h(qs[2])
    h(qs[2])
    with control(qs[0], qs[1]):
        x(qs[2])
    h(qs[2])
    state_output("out", qs)
    discard_array(qs)

broken.check(); workaround.check()
r1 = broken.with_minimal_opt().emulator(n_qubits=3).statevector_sim().with_seed(0).run()
r2 = workaround.with_minimal_opt().emulator(n_qubits=3).statevector_sim().with_seed(0).run()
print(r1.partial_state_dicts()[0]["out"].as_single_state())
print(r2.partial_state_dicts()[0]["out"].as_single_state())
```

**Comment by nicolaassolini-qntm** [comment]
@kkoci the issue should already be solved. Try to update`tket` to 0.15.7 version
